retrieval/discovery/ecs.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/aws/aws-sdk-go/service/ec2"
+	ecs "github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"golang.org/x/net/context"
+	"time"
+)
+
+// ECSDiscovery periodically performs ECS-SD requests. It implements
+// the TargetProvider interface.
+type ECSDiscovery struct {
+	aws         *aws.Config
+	interval    time.Duration
+	portFrom    int64
+	portTo      int64
+	clusterName string
+}
+
+// NewECSDiscovery returns a new ECSDiscovery which periodically refreshes its targets.
+func NewECSDiscovery(conf *config.ECSSDConfig) *ECSDiscovery {
+	awsConfig := defaults.Get().Config
+
+	// override the default region
+	if conf.Region != "" {
+		awsConfig.Region = &conf.Region
+	}
+	if conf.AccessKey != "" && conf.SecretKey != "" {
+		creds := credentials.NewStaticCredentials(conf.AccessKey, conf.SecretKey, "")
+		awsConfig.Credentials = creds
+	}
+
+	return &ECSDiscovery{
+		aws:         awsConfig,
+		interval:    time.Duration(conf.RefreshInterval),
+		portFrom:    int64(conf.PortFrom),
+		portTo:      int64(conf.PortTo),
+		clusterName: conf.ClusterName,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ed *ECSDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ed.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := ed.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ed.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (ed *ECSDiscovery) refresh() (*config.TargetGroup, error) {
+	tg := &config.TargetGroup{
+		Source: *ed.aws.Region,
+	}
+	ecsService := ecs.New(session.New(), ed.aws)
+	ec2s := ec2.New(session.New(), ed.aws)
+
+	listServicesResponse, err := ecsService.ListServices(&ecs.ListServicesInput{
+		Cluster:    aws.String(ed.clusterName),
+		MaxResults: aws.Int64(100),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing ecs services: %s", err)
+	}
+
+	services, err := ecsService.DescribeServices(&ecs.DescribeServicesInput{
+		Cluster:  aws.String("default"),
+		Services: listServicesResponse.ServiceArns,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing describing services: %s", err)
+	}
+
+	for _, service := range services.Services {
+		// Possibly cache theses, since ARNS are immutable, afaik
+		taskDefinitionResponse, err := ecsService.DescribeTaskDefinition(&ecs.DescribeTaskDefinitionInput{
+			TaskDefinition: service.TaskDefinition,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error listing describing task description: %s", err)
+		}
+
+		monitored := false
+		monitoringPort := 0
+
+		var dockerLabels map[string]*string
+
+		// See if any container is exposing the specified ports
```

Comment:
How about we have Port be Ports, and an array of numbers. That way, if we detect a task with any of them mapped to the host, we say it's a scrape target. That is the most specific and flexible way probably. Then a single scrape config can handle multiple types of containers. 

That also allows for multiple targets to come from a single service, in case it is a multi-container task, each task with a metrics endpoint.